### PR TITLE
Feature/normalize input

### PR DESCRIPTION
**Pull request by PReichetanz**
Add functionality to normalize name input before saving to localStorage

- whitespace will be removed
- first letter uppercase, all other letters lowercase
- name is saved as first, middle, and last name
- Update Class- and PupilOverview to show the name accordingly

[View at heroku](https://capstone-rev-feature-no-q8ytkq.herokuapp.com/)

### CODE REVIEW

File: src/app/components/Card/Card.tsx [hunk 1]
```diff
@@ -18,7 +18,7 @@ export default function Card({ pupil, deleteCard }: CardProps): JSX.Element {
         handleClick={() => deleteCard(pupil.id)}
       />
       <CardLink to={`/pupil/${pupil.id}`}>
-        <Heading>{name}</Heading>
+        <Heading>{`${name.first} ${name.middle} ${name.last}`}</Heading>
```
Comment (F-Kirchhoff) [inline src/app/components/Card/Card.tsx:21]: I like the naming of the name props

File: src/app/components/utils/normalizeInput.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      last: lastName !== '' ? normalizeInput(lastName) : '',
+    },
+    category: category,
+    evaluation: evaluation,
+  };
+  return normalizedPupil;
+}
+
+function prepareNormalisation(name: string) {
+  return name.trim().replace(/\s+/g, ' ').split(' ');
```
Comment (F-Kirchhoff) [inline src/app/components/utils/normalizeInput.ts:28]: maybe you can give us a short workshop on regex? :D

File: src/app/components/utils/normalizeInput.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import type { NormalizedInput, RawInput } from '../../types/types';
+
+export default function handleInput(pupil: RawInput): NormalizedInput {
+  const preparedName = prepareNormalisation(pupil.name);
+  const firstName = preparedName[0];
+  const middleName =
+    preparedName.length < 2 ? '' : preparedName.slice(1, -1).join(' ');
```
Comment (F-Kirchhoff) [inline src/app/components/utils/normalizeInput.ts]: maybe for a more semantic code you could set the middle name to none if its not there

File: src/app/components/utils/normalizeInput.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  const lastName =
+    preparedName.length === 1
+      ? ''
```
Comment (F-Kirchhoff) [inline src/app/components/utils/normalizeInput.ts]: same here

File: src/app/components/utils/normalizeInput.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      : preparedName[preparedName.length - 1].toLowerCase();
+
+  const category = pupil.category;
+  const evaluation = pupil.evaluation;
```
Comment (F-Kirchhoff) [inline src/app/components/utils/normalizeInput.ts]: you can use the destructuring here: const {category, evaluation } = pupil

File: src/app/components/utils/normalizeInput.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    const replacedWhitespace = firstLetterUpperCase.replace(
+      firstLetterUpperCase[indexOfHyphen - 1],
+      '-'
+    );
+
+    inputToNormalize = replacedWhitespace;
+    return inputToNormalize;
+  }
+  const normalizedInput = toUpperCaseWhitespace(inputToNormalize, 0);
+  return normalizedInput;
```
Comment (F-Kirchhoff) [inline src/app/components/utils/normalizeInput.ts]: just use: return toUpperCaseWhitespace(inputToNormalize, 0);

**Review by F-Kirchhoff — APPROVED** [review]
lgtm :) and really complex too. Had some ideas for further improvement.

**Review by derjubu — APPROVED** [review]
LGTM 🚀

**Pull request merged**